src/util/result.h

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include <iostream>
+
+#include "llvm/Support/Error.h"
+#include "llvm/Support/ErrorOr.h"
+#include "llvm/Support/raw_os_ostream.h"
+
+namespace bpftrace {
+
+// OK is used in order to indicate success without a value type.
+//
+// It is effectively a monostate class, but using this instead of the default
+// llvm `ErrorSuccess` or `Error` means that the bool operator is consistent.
+//
+// To indicate success, you can merely do `return OK()`.
+class OK {};
+
+// Despite the name, in LLVM `ErrorOr` corresponds to a `std::error_code`
+// whilst `Expected` corresponds to a richer LLVM `Error`. The latter is
+// broadly preferred for newer code, as it allows a much richer expression of
+// error information. In an attempt to avoid confusion, we standardize on only
+// the modern versions, but call it `Result` to make its purpose clear. In the
+// future, this could map to a newer standard, such as the standard library
+// `std::expected<T, Error>`, independently of the names in this namespace. The
+// purpose of this file is to approximately insulate against such changes (but
+// not to provide a perfect abstraction, a little bit of leakage is okay).
+//
+// For now, the LLVM `Error` class provides mandatory checking (errors cannot
+// be ignored, they must be explicitly forward or consumed), which is useful.
+//
+// You can use Result as follows:
+//
+//     Result<> foo() {
+//       return OK();
+//     }
+//
+//     Result<Value> bar() {
+//       return Value();
+//     }
+//
+template <typename T = OK>
+using Result = llvm::Expected<T>;
+
+// For error types, one can define custom error type by inheriting from
+// `ErrorInfo`. For example, you may construct an error as follows:
+//
+//    class BpfVerifierError : public ErrorInfo<BpfVerifierError> {
+//    public:
+//      static char ID;
```

Comment:
In case anyone was wondering: this snippet from llvm source explains why this exists. It's quite clever IMO:
```
template <typename ThisErrT, typename ParentErrT = ErrorInfoBase>
class ErrorInfo : public ParentErrT {
public:
  using ParentErrT::ParentErrT; // inherit constructors

  static const void *classID() { return &ThisErrT::ID; }

  const void *dynamicClassID() const override { return &ThisErrT::ID; }

  bool isA(const void *const ClassID) const override {
    return ClassID == classID() || ParentErrT::isA(ClassID);
  }
};
```

Some kind of light weight RTTI